src/packagesystem.rs

```diff
@@ -67,12 +69,182 @@ where
     rpm_parse_metadata(&rpmout.stdout)
 }
 
-#[test]
-fn test_parse_rpmout() {
-    let testdata = "grub2-efi-x64-1:2.06-95.fc38.x86_64,1681321788 grub2-efi-x64-1:2.06-95.fc38.x86_64,1681321788 shim-x64-15.6-2.x86_64,1657222566 shim-x64-15.6-2.x86_64,1657222566 shim-x64-15.6-2.x86_64,1657222566";
-    let parsed = rpm_parse_metadata(testdata.as_bytes()).unwrap();
-    assert_eq!(
-        parsed.version,
-        "grub2-efi-x64-1:2.06-95.fc38.x86_64,shim-x64-15.6-2.x86_64"
-    );
+fn parse_evr_map(input: &str) -> BTreeMap<String, Evr> {
+    input
+        .split(',')
+        .map(|pkg| {
+            // assume it is like "grub2-1:2.12-28.fc42,shim-15.8-3"
+            if !pkg.ends_with(std::env::consts::ARCH) {
+                let (name, version_release) = pkg.split_once('-').unwrap_or((pkg, ""));
+                let evr = Evr::parse(version_release);
+                return (name.to_string(), evr);
+            }
+
+            // assume it is like "grub2-efi-x64-1:2.12-28.fc42.x86_64,shim-x64-15.8-3.x86_64"
+            let nevra = rpm::Nevra::parse(pkg);
+            let _name = nevra.name();
+            // get name as "grub2" to match the usr/lib/efi path
+            let (name, _) = _name.split_once('-').unwrap_or((_name, ""));
+            (
+                name.to_string(),
+                Evr::new(
+                    nevra.epoch().to_string(),
+                    nevra.version().to_string(),
+                    nevra.release().to_string(),
+                ),
+            )
+        })
+        .collect()
+}
+
+// Implement Compare package versions function:
+// Return `Greater` if evr_a > evr_b,
+// Return `Less` if evr_a < evr_b,
+// Return `Equal` if evr_a == evr_b.
+fn compare_package_versions_impl(a: &str, b: &str) -> Vec<(String, Ordering)> {
+    let map_a = parse_evr_map(a);
+    let map_b = parse_evr_map(b);
+
+    // Compare package versions between `map_a` (current) and `map_b` (target).
+    // For each package in `map_b`:
+    // - If it exists in `map_a`, compare their versions and record the result.
+    // - If not found in `map_a`, assume it's a new package that should be upgraded.
+    let mut result = Vec::new();
+    for (name, evr_b) in map_b {
+        if let Some(evr_a) = map_a.get(&name) {
+            let cmp = evr_a.cmp(&evr_b);
+            result.push((name, cmp));
+        } else {
+            log::trace!("Found new package {name} in the target");
+            result.push((name, Ordering::Less));
+        }
+    }
+    result
+}
+
+// Compare package versions:
+// If any package is Ordering::Less, return Ordering::Less, means upgradable,
+// Else if any package is Ordering::Greater, return Ordering::Greater,
+// Else (all equal), return Ordering::Equal.
+pub(crate) fn compare_package_versions(a: &str, b: &str) -> Ordering {
```

Comment:
I think this is OK but it'd be more idiomatic to create a `struct Package` that has `impl Ord`.

And also I think per above we should split on the `,` outside this function, i.e. we're comparing two slices `[Package]` - and if everything compares equal then the longer one wins.

Reply:
Update in https://github.com/coreos/bootupd/pull/983